lib/pbench/test/unit/agent/test_tool_meister.py

```diff
@@ -0,0 +1,202 @@
+
+    @staticmethod
+    def fake_create_tar(returncode: int, stdout: bytes):
+        def f(directory: pathlib.Path, tar_file: pathlib.Path):
+            return subprocess.CompletedProcess(
+                args=[], returncode=returncode, stdout=stdout, stderr=None
+            )
+
+        __class__.functions_called.append("fake_create_tar")
+        return f
+
+    @responses.activate
+    def test_tar_create_success(self, tool_meister, monkeypatch):
+        """This test should pass the tar creation in send directory"""
+
+        def fake_unlink(*args):
+            assert args[0] == pathlib.Path(f"{self.directory}.tar.xz")
+            self.functions_called.append("fake_unlink")
+
+        def fake_open(*args):
+            assert args[0] == pathlib.Path(f"{self.directory}.tar.xz")
+            self.functions_called.append("fake_open")
+            return io.StringIO()
+
+        def fake_rmtree(directory: pathlib.Path):
+            assert directory == tmp_dir
+            self.functions_called.append("fake_rmtree")
+
+        def fake_md5(tar_file: pathlib.Path):
+            assert tar_file == pathlib.Path(f"{self.directory}.tar.xz")
+            self.functions_called.append("fake_md5")
+            return 10, "random_md5"
+
+        monkeypatch.setattr(shutil, "rmtree", fake_rmtree)
+        monkeypatch.setattr("pbench.agent.tool_meister.md5sum", fake_md5)
+        monkeypatch.setattr(pathlib.Path, "unlink", fake_unlink)
+        monkeypatch.setattr(pathlib.Path, "open", fake_open)
+
+        monkeypatch.setattr(tool_meister, "_create_tar", self.fake_create_tar(0, b""))
+
+        url = (
+            f"http://{tm_params['tds_hostname']}:{tm_params['tds_port']}/uri"
+            f"/ctx/{tm_params['hostname']}"
+        )
+        responses.add(responses.PUT, url, status=HTTPStatus.OK, body="succeeded")
+
+        failures = tool_meister._send_directory(self.directory, "uri", "ctx")
+        functions_called, self.functions_called = self.functions_called, []
+        assert functions_called == [
+            "fake_create_tar",
+            "fake_md5",
+            "fake_open",
+            "fake_rmtree",
+            "fake_unlink",
+        ]
+        assert failures == 0
+
+    def test_tar_create_failure(self, tool_meister, monkeypatch):
+        """Check if the tar creation error is properly captured in send_directory"""
+        monkeypatch.setattr(
+            tool_meister,
+            "_create_tar",
+            self.fake_create_tar(1, b"Error in tarball creation"),
+        )
+
+        with pytest.raises(ToolMeisterError) as exc:
+            failures = tool_meister._send_directory(self.directory, "uri", "ctx")
+            assert failures == 1
+        functions_called, self.functions_called = self.functions_called, []
+        assert functions_called == ["fake_create_tar"]
+        assert f"Failed to create an empty tar {self.directory}.tar.xz" in str(
+            exc.value
```

Comment:
Once the mock for `_create_tar()` is fixed to actually record the calls to the function, I don't think that this assertion will be necessary (that is, from previous assertions, we'll already know that there were two calls to `_create_tar()` and that the `ToolMeisterError` was raised, so we don't need to also confirm the contents of the log, which we probably shouldn't be doing in this case, anyway).

Reply:
Contents of the log or the exception?

Reply:
Whoops!  My bad...I was thinking that we were looking at the log output, not the exception message.

I'm not sure I care whether we test the exception message or not (especially given that we've already check that we received the exception).